dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java

```diff
@@ -405,4 +410,57 @@ public static long countFailures(List<ArtifactResults> table, String columnName)
         .filter(row -> row.getResult(columnName) == null || row.getFailureCount(columnName) > 0)
         .count();
   }
+
+  private static final int SUMMARIZING_DEPENDENCY_PATH_SIZE_THRESHOLD = 5;
+
+  /**
+   * Returns mapping from jar files to summaries of the root problem in their {@link
+   * DependencyPath}s. The summary explains common patterns ({@code groupId:artifactId}) in the path
+   * elements. The returned map does not have a key for a jar file when the length of its {@link
+   * DependencyPath} list is smaller than {@link #SUMMARIZING_DEPENDENCY_PATH_SIZE_THRESHOLD} or a
+   * common pattern is not found among the elements in the paths.
+   *
+   * <p>Using this summary in the BOM dashboard avoids repetitive items in the {@link
+   * DependencyPath} list that share the same root problem caused by widely-used libraries, for
+   * example, {@code commons-logging:commons-logging}, {@code
+   * com.google.http-client:google-http-client} and {@code log4j:log4j}.
+   */
+  private static ImmutableMap<String, String> findRootCauses(
+      ListMultimap<Path, DependencyPath> jarToDependencyPaths) {
+    // Freemarker is not good at handling non-string key. Path object in .ftl is automatically
+    // converted to String. https://freemarker.apache.org/docs/app_faq.html#faq_nonstring_keys
+    ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
+
+    for (Path jar : jarToDependencyPaths.keySet()) {
+      Set<String> versionlessCoordinatesIntersection = null; // null for 1st iteration
+      List<DependencyPath> dependencyPaths = jarToDependencyPaths.get(jar);
+
+      for (DependencyPath dependencyPath : dependencyPaths) {
+        // List of versionless coordinates ("groupId:artifactId")
+        ImmutableList<String> versionlessCoordinatesInPath =
+            dependencyPath.getPath().stream().map(Artifacts::makeKey).collect(toImmutableList());
+        if (versionlessCoordinatesIntersection == null) {
+          // For 1st iteration, initializes the intersection set with LinkedHashSet, which keeps
+          // the order of the common artifacts.
+          versionlessCoordinatesIntersection = Sets.newLinkedHashSet(versionlessCoordinatesInPath);
+        } else {
+          // intersection of elements in DependencyPaths
+          versionlessCoordinatesIntersection.retainAll(versionlessCoordinatesInPath);
+        }
+      }
+
+      // When dependencyPaths is not empty, versionlessCoordinates is not null
+      if (dependencyPaths.size() > SUMMARIZING_DEPENDENCY_PATH_SIZE_THRESHOLD
+          && versionlessCoordinatesIntersection.size() > 1) { // The last paths elements are always same
```

Comment:
I think the body of this if block could also be pulled into a helper method that returns a string.

Reply:
Changed it to `summaryMessage`